text/0000-rust-llm.md

```diff
@@ -0,0 +1,78 @@
+# RFC: Add large language models to Rust
+
+This adds a new module to the standard library containing interfaces to large language models.
+
+The author highlights that no large language models were used in the making of this RFC exactly because Rust does not support them today.
+
+## Motivation
+
+Large language models (LLMs for short) have become ubiquitous.
+But using AI language models in Rust has been a lot harder than one would expect.
+The standard library provides neither an interface to nor an implementation of a large language model.
+This will significantly hurt adoption in the future as Rust continues to fall behind other ecosystems, like NodeJS built-in `node:llm` or C23's `llm.h`.
+To ensure continued adoption of Rust in enterprise contexts, this RFC proposes enhancing Rust by adding first-class large language model capabilities to its standard library.
+
+## Guide-level explanation
+
+This RFC contains the following changes
+
+- add `std::llm`, containing several interfaces and potentially an implementation of a large language model
+- add a `cargo llm` subcommand for interfacing with `std::llm` from the command line.
+
+The precise API of the standard library model has not yet been defined and should be evaluated by a relevant committee.
+We propose repurposing the existing WG-llvm for this, as it is lexicographically the closest.
+LLVM backend maintenance will be carried out by a large language model in the meantime.
+WG-llvm is believed to have enough capacity for this task, but this should be confirmed with the working group before accepting this RFC.
+
+## Reference-level explanation
+
+The standard library is enhanced by an additional module, `std::llm`.
+This module is part of the `std` crate and not available in `#![no_std]`, as it requires interfacing with an operating system (the large language model, which is basically an operating system).
```

Comment:
I propose splitting `std::llm` into it's own crate called `ai`, and just like `alloc` require the standard library of each platform to provide a default `#[global_llm]` using the built in LLM of the operating system. This would make it easy to use OS-specific features of the LLM, like the copilot button on all laptops produced after 2023 or the kernel level AI in upcoming operating systems like [AIOS](https://arxiv.org/abs/2403.16971). This has the additional benefit of not having to bundle an entire LLM into every produced binary. I estimate this would result in around a 99.95% decrease in binary size for simple applications which is comparable to linking to `sprintf` instead of `core::fmt`.

This has the obvious drawback of not being possible to implement on outdated operating systems like Ubuntu or macOS. The two main ways I can think of to resolve this would be to either follow the precedent from `wasm32-unknown-unknown` and provide a `#[global_llm]` that panics on every call (including the implicit call to initialize the LLM before calling `main`), or to demote all targets except Windows and AIOS to tier 3 support.